event-exporter/kubernetes/podlabels/cache_collector.go

```diff
@@ -0,0 +1,159 @@
+package podlabels
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"golang.org/x/time/rate"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	getPodErrLogRateLimiter = rate.NewLimiter(rate.Every(10*time.Second), 1)
+)
+
+type cacheKey struct {
+	Namespace string
+	Name      string
+}
+
+// PodLabelCollector defines the interface for GetLabels for a namespace and pod name pair.
+type PodLabelCollector interface {
+	GetLabels(namespace, pod string) map[string]string
+}
+
+// podLabelCollectorWithCache implements PodLabelCollector interface with an internal pod label cache.
+type podLabelCollectorWithCache struct {
+	ignoredNamespaces     map[string]struct{}
+	client                kubernetes.Interface
+	getPodTimeout         time.Duration
+	cache                 *lru.Cache[cacheKey, map[string]string]
+	emptyLabelPodCache    *lru.Cache[cacheKey, time.Time] // cache the timestamp so we can check TTL.
+	emptyLabelPodCacheTTL time.Duration
+}
+
+// NewCollector returns a new podLabelCollectorWithCache that maintains a set of (podName, namespace) -> (podLabels) mappings.
+func NewCollector(client kubernetes.Interface, ignoredNamespaces []string, podCacheSize, emptyLabelPodCacheSize int, emptyLabelPodCacheTTL, getPodTimeout time.Duration) (*podLabelCollectorWithCache, error) {
+	cache, err := lru.NewWithEvict[cacheKey, map[string]string](podCacheSize, recordEviction)
+	if err != nil {
+		return nil, err
+	}
+	emptyLabelPodCache, err := lru.NewWithEvict[cacheKey, time.Time](emptyLabelPodCacheSize, recordEmptyLabelPodCacheEvict)
+	if err != nil {
+		return nil, err
+	}
+	pc := &podLabelCollectorWithCache{
+		client:                client,
+		getPodTimeout:         getPodTimeout,
+		ignoredNamespaces:     make(map[string]struct{}),
+		cache:                 cache,
+		emptyLabelPodCache:    emptyLabelPodCache,
+		emptyLabelPodCacheTTL: emptyLabelPodCacheTTL,
+	}
+	for _, ns := range ignoredNamespaces {
+		pc.ignoredNamespaces[ns] = struct{}{}
+	}
+
+	return pc, nil
+}
+
+// Get returns the label mappings for a given Pod. If there's no local cache, we will attempt a
+// get from apiserver.
+func (pc *podLabelCollectorWithCache) GetLabels(namespaceName, podName string) map[string]string {
+	if _, ok := pc.ignoredNamespaces[namespaceName]; ok {
+		return nil
+	}
+	if labels, ok := pc.cache.Get(cacheKey{Namespace: namespaceName, Name: podName}); ok {
```

Comment:
I noticed the recreation of `cacheKey{Namespace: namespaceName, Name: podName}` struct many times in this function, maybe we can define it as a variable?

Reply:
Done.